### DISCUSSION TITLE
What would I need to do to have LiteOs .bin programs run without boot.bin

### DISCUSSION THREAD

**Discussion by JamesTimothyMeech** [BrunoLevy/learn-fpga#118] [Q&A] [answered]
I am using a soft processor running on the Digilent Arty which I generated using this LiteX command: 
```
python3 -m litex_boards.targets.digilent_arty --cpu-type femtorv --cpu-variant gracilis --variant a7-100 --toolchain vivado --with-xadc --with-spi-sdcard --sdcard-adapter digilent --timer-uptime --bios-format double --build --load
```
I want to be able to send binaries directly to the soft processor on the Arty without the need to unplug the SD card from the Arty and into my laptop to copy the binary files across. When I run 
```
james@james-Laptop-13th-Gen-Intel-Core:~/Desktop/Casino/FPGA-System-on-Chip-Firmware/LiteX/software/Programs$ litex_term /dev/ttyUSB1 --speed 115200 --kernel double_add.bin 

        __   _ __      _  __
       / /  (_) /____ | |/_/
      / /__/ / __/ -_)>  <
     /____/_/\__/\__/_/|_|
   Build your hardware, easily!

 (c) Copyright 2012-2024 Enjoy-Digital
 (c) Copyright 2007-2015 M-Labs

 BIOS built on Jan 28 2024 00:04:40
 BIOS CRC passed (e61ca211)

 LiteX git sha1: 245bed71

--=============== SoC ==================--
CPU:		FemtoRV-GRACILIS @ 100MHz
BUS:		WISHBONE 32-bit @ 4GiB
CSR:		32-bit data
ROM:		128.0KiB
SRAM:		8.0KiB
L2:		8.0KiB
SDRAM:		256.0MiB 16-bit @ 800MT/s (CL-7 CWL-5)
MAIN-RAM:	256.0MiB

--========== Initialization ============--
Initializing SDRAM @0x40000000...
Switching SDRAM to software control.
Read leveling:
  m0, b00: |00000000000000000000000000000000| delays: -
  m0, b01: |00000000000000000000000000000000| delays: -
  m0, b02: |11111111111000000000000000000000| delays: 05+-05
  m0, b03: |00000000000000111111111111110000| delays: 20+-06
  m0, b04: |00000000000000000000000000000001| delays: -
  m0, b05: |00000000000000000000000000000000| delays: -
  m0, b06: |00000000000000000000000000000000| delays: -
  m0, b07: |00000000000000000000000000000000| delays: -
  best: m0, b03 delays: 20+-06
  m1, b00: |00000000000000000000000000000000| delays: -
  m1, b01: |00000000000000000000000000000000| delays: -
  m1, b02: |11111111111100000000000000000000| delays: 05+-05
  m1, b03: |00000000000000011111111111110000| delays: 21+-06
  m1, b04: |00000000000000000000000000000000| delays: -
  m1, b05: |00000000000000000000000000000000| delays: -
  m1, b06: |00000000000000000000000000000000| delays: -
  m1, b07: |00000000000000000000000000000000| delays: -
  best: m1, b03 delays: 21+-06
Switching SDRAM to hardware control.
Memtest at 0x40000000 (2.0MiB)...
  Write: 0x40000000-0x40200000 2.0MiB     
   Read: 0x40000000-0x40200000 2.0MiB     
Memtest OK
Memspeed at 0x40000000 (Sequential, 2.0MiB)...
  Write speed: 25.3MiB/s
   Read speed: 35.9MiB/s

--============== Boot ==================--
Booting from serial...
Press Q or ESC to abort boot completely.
sL5DdSMmkekro
[LITEX-TERM] Received firmware download request from the device.
[LITEX-TERM] Uploading double_add.bin to 0x40000000 (70144 bytes)...
[LITEX-TERM] Upload calibration... (inter-frame: 10.00us, length: 64)
[LITEX-TERM] Upload complete (9.9KB/s).
[LITEX-TERM] Booting the device.
[LITEX-TERM] Done.
Executing booted program at 0x40000000

--============= Liftoff! ===============--
```
in an attempt to directly run a LiteOS program without using boot.bin my program doesn't run. What am I missing here? I can see that I can potentially batch files together using a .json description file: https://github.com/enjoy-digital/litex/wiki/Load-Application-Code-To-CPU but I am not sure how to calculate the memory locations that go alongside the file names.

**Comment by BrunoLevy** [comment, marked as answer]
Hi,

The FemtOS elf binaries and the binaries that you can send directly to LiteX through litex_term have a different memory map (different linker script, different makefiles).

If you want to compile a program that you can directly send to litex_term, you can use [this one](https://github.com/BrunoLevy/learn-fpga/tree/master/LiteX/software/DemoBundle) as an example (in addition, it is interfaced with LiteX command line facilities, so that you can have several functionalities in your firmware, like different commands). It is what I was using before developing FemtOS.

Another option is using a ULX3S: it has an ESP32 that can be used to directly transfer ELF binaries on the SDCard.

**Reply by JamesTimothyMeech** [reply]
Thanks! It works for me on the Arty, using the ULX3S is not an option because I am using the XADC.

**Reply by JamesTimothyMeech** [reply]
Is there an easy way to have LiteOS run a program over and over again by having my PC send commands over the terminal? I have been trying to use the unix expect utility to do this but so far all I have managed to do is create a program that opens the LiteX terminal and then does nothing:
```
#!/usr/bin/expect

spawn litex_term /dev/ttyUSB1 --speed 115200
set prompt "liteOS>"
expect "liteOS>\r"
send "run double_add_print.elf\n"
```

**Reply by BrunoLevy** [reply]
Hi James,

I'm pretty sure it is possible to do,

I'm unsure how, I'd suggest trying to send a carriage return before expecting the prompt.

**Reply by JamesTimothyMeech** [reply]
Thanks for getting back to me! For now I am putting the entire program inside a for loop as a workaround.

**Reply by BrunoLevy** [reply]
From your previous script, I guess it waits for the 'LiteOS>' prompt forever, because it is probably sent *before* the script expects it. If you send a carriage return, then LiteOS will send the prompt, you can expect it, then send the command you want to run. I am completely unsure of what I'm saying, that's just a theory !
